pkg/volume/csi/csi_test.go

```diff
@@ -0,0 +1,423 @@
+			},
+			isInline: true,
+		},
+		{
+			name:     "missing spec",
+			specName: "pv2",
+			driver:   "simple-driver",
+			volName:  "vol2",
+			specFunc: func(specName, driver, volName string) *volume.Spec {
+				return nil
+			},
+			podFunc: func() *api.Pod {
+				podUID := types.UID(fmt.Sprintf("%08X", rand.Uint64()))
+				return &api.Pod{ObjectMeta: meta.ObjectMeta{UID: podUID, Namespace: testns}}
+			},
+			shouldFail: true,
+		},
+		{
+			name:     "incompete spec",
+			specName: "pv2",
+			driver:   "simple-driver",
+			volName:  "vol2",
+			specFunc: func(specName, driver, volName string) *volume.Spec {
+				return &volume.Spec{ReadOnly: true}
+			},
+			podFunc: func() *api.Pod {
+				podUID := types.UID(fmt.Sprintf("%08X", rand.Uint64()))
+				return &api.Pod{ObjectMeta: meta.ObjectMeta{UID: podUID, Namespace: testns}}
+			},
+			shouldFail: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tmpDir, err := utiltesting.MkTmpdir("csi-test")
+			if err != nil {
+				t.Fatalf("can't create temp dir: %v", err)
+			}
+			defer os.RemoveAll(tmpDir)
+			client := fakeclient.NewSimpleClientset()
+			fakeWatcher := watch.NewRaceFreeFake()
+
+			factory := informers.NewSharedInformerFactory(client, csiResyncPeriod)
+			go factory.Start(wait.NeverStop)
+
+			host := volumetest.NewFakeVolumeHostWithCSINodeName(
+				tmpDir,
+				client,
+				nil,
+				"csi-node",
+				factory.Storage().V1beta1().CSIDrivers().Lister(),
+			)
+
+			plugMgr := &volume.VolumePluginMgr{}
+			plugMgr.InitPlugins(ProbeVolumePlugins(), nil /* prober */, host)
+			csiClient := setupClient(t, true)
+
+			volSpec := test.specFunc(test.specName, test.driver, test.volName)
+			pod := test.podFunc()
+			attachName := getAttachmentName(test.volName, test.driver, string(host.GetNodeName()))
+			t.Log("csiTest.VolumeAll starting...")
+
+			// *************** Attach/Mount volume resources ****************//
+			// attach volume
+			t.Log("csiTest.VolumeAll Attaching volume...")
+			attachPlug, err := plugMgr.FindAttachablePluginBySpec(volSpec)
+			if err != nil {
+				if !test.shouldFail {
+					t.Fatalf("csiTest.VolumeAll PluginManager.FindAttachablePluginBySpec failed: %v", err)
+				} else {
+					t.Log("csiTest.VolumeAll failed: ", err)
+					return
+				}
+			}
+
+			if test.isInline && attachPlug != nil {
+				t.Fatal("csiTest.VolumeAll AttachablePlugin found with ephemeral volume")
+			}
+			if !test.isInline && attachPlug == nil {
+				t.Fatal("csiTest.VolumeAll AttachablePlugin not found with PV")
+			}
+
+			var devicePath string
+			if attachPlug != nil {
+				t.Log("csiTest.VolumeAll attacher.Attach starting")
+
+				var volAttacher volume.Attacher
+
+				volAttacher, err := attachPlug.NewAttacher()
+				if err != nil {
+					t.Fatal("csiTest.VolumeAll failed to create new attacher: ", err)
+				}
+
+				go func(spec *volume.Spec, nodeName types.NodeName) {
```

Comment:
What's the purpose of running Attach in a go routine?

Reply:
It is to avoid a deadlock in the test since CSI attacher immediately starts waiting for the status of `VolumeAttachment` to be updated to true.  Both cant happen on the same go routine.

Reply:
can you add that as a comment?

Reply:
Added comment.